_posts/2018-12-04-imposing-invariants-in-scala.md

```diff
@@ -0,0 +1,353 @@
+We also have:  
+
+```scala
+case class Ticket[SeatType <: Seat](
+  verificationCode: String,
+  price: Int,
+  seat: SeatType
+)
+```
+
+```scala
+case class Reservation[SeatType <: Seat](
+  userTicket: Ticket[SeatType],
+  guestTicket: Ticket[SeatType]
+)
+```
+
+
+The first thing to notice is that the above does not deliver what we want. The following compiles:  
+
+```scala
+val reservation = Reservation(
+  usertTicket = Ticket("id1", 123, VipTribuneSeat("east")),
+  guestTicket = Ticket("id2", 123, Standing)
+)
+```
+
+The compiler tries to infer the type <code>SeatType</code>for reservation. From the user ticket on line 2 it thinks it is <code>VipTribuneSeat</code>. The compiler is satisfied since this a sub-type of <code>Seat</code> which is an upper type bound on Reservation. This is contradicted on the 3rd line, as we encounter type <code>Standing</code> for the guest ticket. Bad news since type <code>SeatType</code> only line 2 and <code>SeatType</code> on line 3 refer, well, to the same type!  
+Before launching an error, the compiler realizes both <code>Standing </code> and <code>VipTribuneSeat</code> are actually of sub-types of <code>Seat</code> ; this means they can pretend to be their super-type (<code>Seat</code>). Because <code>Seat</code> respects the type bound <code>Seat</code> <:  <code>Seat</code>, the compilation is successful with val reservation: Reservation[Seat].
+
+<strong> There are at least 4 ways to enforce the restriction of the opera-house: </strong>
+
+
+1. using require(condition) on the class constructor of Reservation  
+2. Hiding the public constructor of the class, defining a public one that validates and returns Option[Reservation]
+3. Using generalized type constrains.
+4. Using OOP - subtyping
+
+
+# Using require(condition) on the class constructor of Reservation
+
+```scala
+case class Reservation[SeatInfo <: Seat](
+  userTicket: Ticket[SeatInfo],
+  guestTicket: Ticket[SeatInfo]
+){
+
+  private def verifyTypeRestriction: Boolean = (userTicket.seat, guestTicket.seat) match {
+    case (NormalSeat(_, _), NormalSeat(_, _)) => true
+    case (VipTribuneSeat(_), VipTribuneSeat(_)) => true
+    case (_: Standing.type , _: Standing.type ) => true
+    case _ => false
+  }
+
+  require(verifyTypeRestriction)
+}
+```
+
+On this approach we delegate to the users of case class Reservation the responsibility of never trying to instantiate a value that does not conform to the rules. If they do breach the rule, the code compiles but throws an exception at run time.  
+
+While this is the least adequate of the solutions, we can the peace of mind that, if it does blow up, it will do so on instantiation of the object. Meaning that some other block of code further down can expect the invariant to be respected and avoid dealing with the inconsistent cases (since if the block of code is being executed, the previous instantiation succeed). We have centralized the point of failure to the instantiation of an object.  
```

Comment:
we can **____** peace of mind
I think `the` does not fit here, try some verb `have` `get` `find`...

Reply:
done.